### PR TITLE
agent: Rework the debug console

### PR DESCRIPTION
**Pull request by Tim-Zhang**
Rework the debug console with `tokio` + async and make it clear and easy to read.

### CODE REVIEW

**Review by liubin — APPROVED** [review]

**Review by jodh-intel — APPROVED** [review]
Thanks @Tim-Zhang!

lgtm!

 I've tested this manually with a variety of config options and agent shutdown works in all of them.

I'll be raising a CI test for shutdown soon to avoid any further agent shutdown issues hopefully.

**Comment by jodh-intel** [comment]
Restarted Debian CI which failed with what looks like a timeout?

```
not ok 1 Parallel jobs
# (in test file k8s-parallel.bats, line 33)
#   `kubectl wait --for=condition=Ready pod -l jobgroup=${job_name}' failed
# INFO: k8s configured to use runtimeclass
# job.batch/process-item-test1 created
# job.batch/process-item-test2 created
# job.batch/process-item-test3 created
# NAME                 COMPLETIONS   DURATION   AGE
# process-item-test1   0/1           0s         1s
# process-item-test2   0/1           0s         0s
# process-item-test3   0/1           0s         0s
# pod/process-item-test2-qbkwq condition met
# pod/process-item-test3-q5sqk condition met
# error: timed out waiting for the condition on pods/process-item-test1-64gqr

     :

{"level":"warn","ts":"2021-04-07T14:27:15.697Z","caller":"clientv3/retry_interceptor.go:61","msg":"retrying of unary invoker failed","target":"endpoint://client-92172e7c-1aef-4b19-a6c6-95dc8403d312/10.0.0.30:2379","attempt":0,"error":"rpc error: code = Unknown desc = etcdserver: re-configuration failed due to not enough started members"}
W0407 14:27:15.697476    4744 removeetcdmember.go:61] [reset] failed to remove etcd member: etcdserver: re-configuration failed due to not enough started members
.Please manually remove this etcd member using etcdctl
```

File: src/agent/src/util.rs
Comment (lifupan) [inline src/agent/src/util.rs:68]: The listenFd had been set as CLOEXEC, why it needed to reset again here?

File: src/agent/src/console.rs
Comment (lifupan) [inline src/agent/src/console.rs:220]: It's better use a "child" var name instead of "cmd" here.

File: src/agent/src/util.rs
Comment (Tim-Zhang) [reply src/agent/src/util.rs:68]: Yes, it was a mistake, will remove it.

File: src/agent/src/console.rs
Comment (Tim-Zhang) [reply src/agent/src/console.rs:220]: ok

File: src/agent/src/util.rs
Comment (Tim-Zhang) [reply src/agent/src/util.rs]: done

File: src/agent/src/console.rs
Comment (Tim-Zhang) [reply src/agent/src/console.rs]: done

File: src/agent/src/console.rs
Comment (mxpv) [inline src/agent/src/console.rs:57]: I don't think we need to clone `Vec` just to search it.
This can be simplified to
```rust
    let shells = SHELLS
        .lock()
        .unwrap()
        .iter()
        .find(|sh| PathBuf::from(sh).exists())
        .ok_or_else(|| anyhow!("no shell found to launch debug console"))?;
```

File: src/agent/src/console.rs
Comment (mxpv) [inline src/agent/src/console.rs:152]:
```suggestion
    let _ = unistd::execvp(cmd.as_c_str(), &[]).map_err(|e| match e {
```

File: src/agent/src/console.rs
Comment (Tim-Zhang) [reply src/agent/src/console.rs:57]: I will simplify it to 
```
    let shells = SHELLS.lock().unwrap().to_vec();

    let shell = shells
        .into_iter()
        .find(|sh| PathBuf::from(sh).exists())
        .ok_or_else(|| anyhow!("no shell found to launch debug console"))?
```

The reason is that we have keep a temp var to hold the lock and we need get the ownership from `to_vec + into_iter`

File: src/agent/src/console.rs
Comment (Tim-Zhang) [reply src/agent/src/console.rs:152]: will do that

Comment (Tim-Zhang) [reply src/agent/src/console.rs]: done

**Comment by fidencio** [comment]
@Tim-Zhang, this PR is ready to be merged but I'd like to get the approval from @mxpv before merging it, thus I'm adding the "do-not-merge" label for now, giving @mxpv the chance to comment on it later Today.

Thanks for the PR and I really want to see this as part of the release we're planning for Tomorrow!

**Comment by Tim-Zhang** [comment]
> Thanks for the PR and I really want to see this as part of the release we're planning for Tomorrow!

ok

File: src/agent/src/console.rs
Comment (Tim-Zhang) [reply src/agent/src/console.rs]: done @mxpv

**Review by mxpv — APPROVED** [review]
LGTM

**Review by lifupan — APPROVED** [review]

**Pull request merged**
